File: nhay.py
def KMPSearch(pat, txt):
    M = len(pat)
    N = len(txt)
    lps = [0]*M
    j = 0
    computeLPSArray(pat, M, lps)

    i = 0
    while i < N:
        if pat[j] == txt[i]:
            i += 1
            j += 1
        if j == M:
            print str(i-j)
            j = lps[j-1]
        # mismatch after j matches
        elif i < N and pat[j] != txt[i]:
            if j != 0:
                j = lps[j-1]
            else:
                i += 1

def computeLPSArray(pat, M, lps):
    lps[0] = 0
    len = 0
    i = 1
    while i < M:
        if pat[i] == pat[len]:
            len += 1
            lps[i] = len
            i += 1
        else:
            if len != 0:
                len = lps[len-1]
            else:
                lps[i] = 0
                i += 1

txt = str(raw_input())
pat = str(raw_input())
KMPSearch(pat, txt)

File: nhay.cpp
#include <bits/stdc++.h>
using namespace std;

void computeLPSArray(string pat, int M, int *lps)
{
    int len = 0;  
    int i;

    lps[0] = 0; 
    i = 1;
    while (i < M)
    {
		if (pat[i] == pat[len])
		{
			len++;
			lps[i] = len;
			i++;
		}
		else
		{
			if (len != 0)
				len = lps[len-1];
	
			else 
			{
				lps[i] = 0;
				i++;
			}
		}
	}
}

void KMPSearch(string pat, string txt)
{
	int M = pat.length();
	int N = txt.length();
	int *lps;
	lps = new int[M];
    int i = 0, j  = 0;
    computeLPSArray(pat, M, lps); 
   
    while (i < N)
    {
		if (pat[j] == txt[i])
		{
			j++;
			i++;
		}
		if (j == M)
		{	
			cout << i-j << endl;
			j = lps[j-1];
		}
		else if (pat[j] != txt[i])
		{
			if(j == 0)
				i++;
			else
				j = lps[j-1];
		}
    }
    delete lps;
}

int main()
{
	string pat, txt;
	int n;
	while(scanf("%d",&n)==1)
	{
		cin >> pat >> txt;
		KMPSearch(pat, txt);
		cout << endl;
	}
   return 0;
}
